cmd/build.go

```diff
@@ -0,0 +1,203 @@
+)
+
+var dockerAPIVersion = "1.23"
+
+var job = "build"
+var nodeTotal = 1
+var checkoutKey = "~/.ssh/id_rsa"
+var skipCheckout = true
+
+func newBuildCommand() *cobra.Command {
+
+	command := &cobra.Command{
+		Use:   "build",
+		Short: "Run a build",
+		RunE:  runBuild,
+	}
+
+	/*
+		TODO: Support these flags:
+				--branch string         Git branch
+			-e, --env -e VAR=VAL        Set environment variables, e.g. -e VAR=VAL
+				--index int             node index of parallelism
+				--repo-url string       Git Url
+				--revision string       Git Revision
+			-v, --volume stringSlice    Volume bind-mounting
+	*/
+
+	flags := command.Flags()
+	flags.StringVarP(&dockerAPIVersion, "docker-api-version", "d", dockerAPIVersion, "The Docker API version to use")
+	flags.StringVar(&checkoutKey, "checkout-key", checkoutKey, "Git Checkout key")
+	flags.StringVar(&job, "job", job, "job to be executed")
+	flags.BoolVar(&skipCheckout, "skip-checkout", skipCheckout, "use local path as-is")
+	flags.IntVar(&nodeTotal, "node-total", nodeTotal, "total number of parallel nodes")
+
+	return command
+}
+
+var picardRepo = "circleci/picard"
+var circleCiDir = path.Join(settings.UserHomeDir(), ".circleci")
+var currentDigestFile = path.Join(circleCiDir, "current_picard_digest")
+var latestDigestFile = path.Join(circleCiDir, "latest_picard_digest")
+
+func getDigest(file string) string {
+
+	if _, err := os.Stat(file); !os.IsNotExist(err) {
+		digest, err := ioutil.ReadFile(file)
+
+		if err != nil {
+			Logger.Error("Could not load digest file", err)
+		} else {
+			return strings.TrimSpace(string(digest))
+		}
+	}
+
+	return "" // Unknown digest
+}
+
+func newPullLatestImageCommand() *exec.Cmd {
+	return exec.Command("docker", "pull", picardRepo)
+}
+
+func getLatestImageSha() (string, error) {
+
+	cmd := newPullLatestImageCommand()
+	Logger.Info("Pulling latest build image")
+	bytes, err := cmd.CombinedOutput()
+
+	if err != nil {
+		return "", errors.Wrap(err, "failed to pull latest image")
+	}
+
+	output := string(bytes)
+
+	sha256 := regexp.MustCompile("(?m)sha256.*$")
+
+	//latest_image=$(docker pull picardRepo | grep -o "sha256.*$")
+	//echo $latest_image > $LATEST_DIGEST_FILE
+	latest := sha256.FindString(output)
+
+	if latest == "" {
+		return latest, errors.New("Failed to find latest image")
+	}
+
+	return latest, nil
+}
+
+func picardImage() (string, error) {
+	currentDigest := getDigest(currentDigestFile)
+
+	if currentDigest == "" {
+		// Receiving latest image of picard in case of there's no current digest stored
+		Logger.Info("Downloading latest CircleCI build agent...")
+
+		var err error
+
+		currentDigest, err = getLatestImageSha()
+
+		// TODO - write the digest to a file so that we can
+		// use it again.
+
+		if err != nil {
+			return "", err
+		}
+
+	} else {
+		// Otherwise pulling latest image in background
+		// TODO - this should write to a file to record
+		// the fact that we have the latest image.
+		newPullLatestImageCommand().Start()
```

Comment:
I think we should reconsider this "pull latest in background on every run". I think it's unexpected behaviour and not a good user experience. This is Windows Updates.

Instead I think we could provide a subcommand to update it.